src/core_plugins/kibana/public/context/app.js

```diff
@@ -0,0 +1,100 @@
+import _ from 'lodash';
+
+import uiModules from 'ui/modules';
+import contextAppTemplate from './app.html';
+import './components/loading_button';
+import './components/size_picker/size_picker';
+import {
+  createInitialQueryParametersState,
+  QueryParameterActionsProvider,
+  QUERY_PARAMETER_KEYS,
+} from './query_parameters';
+import { QueryActionsProvider } from './query';
+
+const module = uiModules.get('apps/context', [
+  'elasticsearch',
+  'kibana',
+  'kibana/config',
+  'kibana/notify',
+  'ngRoute',
+]);
+
+module.directive('contextApp', function ContextApp() {
+  return {
+    bindToController: true,
+    controller: ContextAppController,
+    controllerAs: 'contextApp',
+    restrict: 'E',
+    scope: {
+      anchorUid: '=',
+      columns: '=',
+      indexPattern: '=',
+      predecessorCount: '=',
+      successorCount: '=',
+      sort: '=',
+      discoverUrl: '=',
+    },
+    template: contextAppTemplate,
+  };
+});
+
+function ContextAppController($scope, Private) {
+  const queryParameterActions = Private(QueryParameterActionsProvider);
+  const queryActions = Private(QueryActionsProvider);
+
+  this.state = createInitialState(this.discoverUrl);
+
+  this.actions = _.mapValues(Object.assign(
+    {},
+    queryParameterActions,
+    queryActions,
+  ), (action) => (...args) => action(this.state)(...args));
+
+  $scope.$watchGroup([
+    () => this.state.rows.predecessors,
+    () => this.state.rows.anchor,
+    () => this.state.rows.successors,
+  ], _.spread(this.actions.setAllRows));
```

Comment:
I brought it up because I had to look up both `_.spread` and the interface for the `$watchGroup` callback and piece them together. 😄 

I think you might be looking at this a little too philosophically... I'm looking at this from the perspective of: "How do we make this code more readable? How do we make it easier to understand?" I'm not saying people _shouldn't_ learn about lodash's API or about `$watchGroup`. But I want to make our code as readable as possible.

As someone who got tripped up on the original code, I find my suggestion easier to understand, because the callback is painfully obvious.  I'm not married to this though... if you have other ideas, I'm open to them. :)